### ISSUE TITLE
Move e2e test setup from Bash to go

### ISSUE THREAD

**Issue by randomvariable** [kubernetes-sigs/cluster-api-provider-vsphere#2044]
/kind feature

**Describe the solution you'd like**

hack/e2e.sh includes a bunch of Node IPAM setup code that would better be served being added to a ginkgo SynchronizedBeforeSuite setup for better debuggability.

CAPA has good examples of using this method to pass data to test suites in https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/main/test/e2e/suites/conformance/conformance_suite_test.go#L45C7-L49 and https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/main/test/e2e/shared/suite.go#L209-L269 

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- Cluster-api-provider-vsphere version: 
- Kubernetes version: (use `kubectl version`): 
- OS (e.g. from `/etc/os-release`):

**Comment by sbueringer** [comment]
Sounds great. Let's just be careful with introducing additional Go dependencies that could lead to problems (e.g. importing some API types which then makes us dependent on a specific CR version)

**Comment by sbueringer** [comment]
To give an example. This api group for example: https://github.com/metal3-io/ip-address-manager/blob/main/api/v1alpha1/ipaddress_webhook.go

But I'm pretty sure we can work around this by e.g. just using Unstructured

**Comment by randomvariable** [comment]
Yeah, that's probably best. There's also some very nice quality of life stuff in the CAPA one, especially the goroutines that captures logs on a ticker that you can run when developing e2e tests and testing code.

**Comment by chrischdi** [comment]
Note: Cloud Provider vSphere seems to use the same infra + script (minus our latest changes): https://github.com/kubernetes/cloud-provider-vsphere/blob/master/hack/e2e.sh

So moving it for CAPV to go would still leave the cloud provider on the old script.

**Comment by sbueringer** [comment]
I think they can probably copy&paste what we do. Basically they do almost the same as we do in e2e tests (they use the CAPI test framework with clusterctl, etc.)

**Comment by randomvariable** [comment]
CPI should really be using a kubetest2 implementation so they can walk the latest recommended builds of everything, like https://github.com/kubernetes-sigs/provider-aws-test-infra/tree/main/kubetest2-ec2

**Comment by chrischdi** [comment]
Note: IPAM for CI moved to Go in capv and is not anymore inside the `e2e.sh` script :-)
